fractal-puppy/Engine
# Feature/ui hover check

When 2 _hoverable_ UI elements overlapped,  both behaviors would trigger on mouse hover instead of just triggering the top element.

- UI module tracks the top ComponentButton hovered (using UI ordering)
- Image is considered hovered if its ordering is above the registered "top button". In this case, both the image and the button will be considered hovered (Only buttons should be used to track hovering but this patch should be functional for any legacy behaviour)
- Added a PreUpdate method to components (allow checking the ordering before component update). 
  - The Component PreUpdate Method runs during the GameObject Update phase.
- Added OnButtonDown() and OnButtonUp() events on Scripts
  - ComponentButton will trigger all OnButtonDown()/OnButtonUp() events for all scripts in the **same gameobject**
- Now, the IsHovered method on buttons check if the button is being hovered and it's on top of the ui order
- the top button is stored in the UI Module.
  - The top hovered button component can be retrieved using GetButtonHover (can return nullptr!!!)
  - A new candidate can be proposed using SetHoverButton (the method will check if the proposed button is on top of the previously registered and modify it if necessary)
  - The stored value gets reset to nullptr every Component PreUpdate

----
**Test:**
- **Recompile all the scripts** (since new method have been added to the basescript and component classes, it will raise errors on previously compiled dlls)
- Open the Graveyard/Temple
- Create a new button and check that current uiOrder below the HUD uiOrder (you can select any image to see the boundaries of the button for the button but "white" would do the trick,)
- Move the newly created button to a location where it overlaps with any HUD element
- Check that the hover behavior is triggered only on the HUD element
- Set the ui order for the new button greater than the ui order for the hud (1000 should do the trick) and check that now only the new button's hover behavior gets triggered

---
![UiHover](https://user-images.githubusercontent.com/26639060/63881182-8ba88080-c9cf-11e9-8a40-284ed0ae9279.gif)
